s3migrate/README.md

```diff
@@ -0,0 +1,154 @@
+- `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`
+- `AWS_S3_REGION_NAME=us-east-1`
+- `SQS_URL` the URL of the SQS queue
+- DYNAMODB_TABLE` The name of the dynamodb table created above.
+- `BUCKETS` a comma separated list of colon based mappings from old bucket names to new ones. For example: `old-bucket:new-bucket,old-bucket2:new-bucket2`. There must be a mapping for every old bucket in mogile that contains data.
+
+### sync data from mogile to an AWS bucket
+
+This step is optional. It is an optimization step to avoid large-scale data transfer from mogile. We have performed this step for production.
+
+The data from our mogile repositories should be moved in place to the S3 bucket. For instance, a file stored in `0/000/000/0000000001.fid` should be stored with the object key `/0/000/000/0000000001.fid` in the S3 bucket.
+
+### Remove or fix bad data
+
+All objects with a dkey of the form `{UUID}.tmp` stored in mogile should be removed. These are temporary files left over from failed contentrepo ingests.
+
+These can be identified in the mogile database using:
+```
+> select dkey from file where dkey like "%.tmp";
+```
+
+The files can be delete from mogile using the python client:
+```
+pipenv run python
+>>> c = pymogilefs.client.Client( trackers=os.environ['MOGILE_TRACKERS'].split(','), domain='plos_repo')
+>>> c.delete_file(DKEY)
+```
+for each `DKEY` identified in the previous step.
+
+All files with known bad checksums stored in mogile using the wrong content addressable storage should be fixed. There is only one known instance of this.
+
+```
+UPDATE file SET dkey='f79ba8f5b9fecbeb8ace12421c73d5198bea6d05-mogilefs-prod-repo' where dmid = 1 and dkey = 'acec11893a476ca42e95c8d546431202ea821210-mogilefs-prod-repo';
+```
+
+### Run a large-scale migration in AWS lambda
+
+The following command should be run:
+```
+pipenv run python enqueue.py migrate
+```
+
+This will enqueue all items in mogile to lambda for processing. These jobs can be monitored using the AWS console.
+
+### Stop ingesting new data into contentrepo
+
+This should happen just before we are ready to switch over to S3.
+
+### Run a large-scale migration in AWS lambda
+
+Rerun the following command:
+```
+pipenv run python enqueue.py migrate
+```
+
+This should only take a few hours, as the vast majority of the items should have been migrated over in a previous step. When the jobs have all succeeded successfully, move on.
+
+### Verify migration
+
+First, verify that the total number of items in the dynamodb table matches the total number of items in mogile. This can be done by running:
+
+```
+$ pipenv run python fids.py > fids.list
+$ wc -l fids.list
+```
+
+and compare with (mysql on mogile)
+```
+> SELECT COUNT(fid) from file;
+```
+
+If these counts do not match, there is a problem.
+
+Next, we can run the AWS lambda based checksum verification:
+```
+pipenv run python enqueue.py verify
+```
+
+All lambda jobs should finish without error.
```

Comment:
How do you check that? Through cloudwatch metrics?

Reply:
Good point. I've been checking the SQS queue: it should completely empty out. If anything is left over, you can check the cloudwatch logs. I've struggle with getting good error reporting from Lambda, because errors are expected and retries are built in. For that reason I also built a non-lambda based verification script, `verify.py`.

Reply:
seems like you have it covered, but another option for getting a list of failed jobs is by setting up an SQS dead letter queue: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-dead-letter-queues.html You could, for instance, tell lambda to retry failed jobs twice (the lambda default), then if they're still failing, pop the job on the dead letter queue. Then you have a nice list of everything that failed.

Reply:
Ooooh, someone that passed me by. I'll add that to the list.